Sammyjo20/Wagonwheel
# Allow mailables to be stored online for an indefinite amount of time

## Introduction
This PR allows mailables to **never expire**, in addition to the currently existing configuration options to specify an amount in days before the mailable expires.

```
<?php

return [
    /*
     * ...
     * To store indefinitely, set this value to: 0.
     */
    'message_expires_in_days' => 0,
];
```

## What has been done
In this PR there are a lot of files changed, since I took an opportunity to refactor some logic around: 
- the dertermination of an expiry date 
- the generation of the signed route URL

I'll list the updates I made below:
- Added explanation (doc block) to store mailables indefinitely in `config/wagonwheel.php`
- Added unit tests for the `OnlineMailable` model
  - Test that a mailable can have a `null` expiration date
- Added test to `DeleteExpiredMailables` command to assure it doesn't delete `OnlineMailables` with a `null` expiration date
- Extract logic for determining the expiration date to the `OnlineMailable` model in a `getExpirationDate()` method.
- Extract logic for building signed route URL's to the `OnlineMailable` model itself in a `getSignedUrl` method.
  - This method returns a `temporarySignedRoute` when a mailable has a expiration date
  - It returns a `signedRoute` (which doesn't expire) when a mailable does not have an expiration date
- Removed redundant `UrlHelper` class
- Updated code to make use of the methods on the `OnlineMailable` model
- Added feature test for the package, testing that: 
  - sent e-mails have an online variant when the `SaveForOnlineViewing` trait is used
  - mailables with expiration date URL's are valid until the expiration date
  - mailables without expiration date are always accessible

## How to test
You can test the functionality within this PR by: 
- checking out this branch 
- adding the package to a Laravel package (locally) ([instructions](https://laravelpackage.com/02-development-environment.html#importing-the-package-locally)) 
- add the `SaveForOnlineViewing` trait to a certain Laravel Mailable
- publish the packages configuration `php artisan vendor:publish --tag=wagonwheel-config`
- change the `'message_expires_in_days'` value to  `0`.
- send the e-mail from the Laravel app (using the log driver or mailtrap of mailhog or ...)
- assert that the database has an `expire_at` column of `null` for the mailable in the `online_mailables` table
- run the command `php artisan wagonwheel:delete-expired-mailables`
- assert that the command does not delete this mailable
- assert that you can view this e-mail in the browser, by using the generated signed URL (check the logs / mailtrap / mailhog /...)

## Notes
It might be well possible that there are other refactors to be made, but I tried to limit refactoring the code to the scope of allowing indefinite availability of online mailables.
